src/System.Net.Sockets/tests/FunctionalTests/Accept.cs

```diff
@@ -288,6 +288,75 @@ public async Task AcceptAsync_MultipleAcceptsThenDispose_AcceptsThrowAfterDispos
                 }
             }
         }
+
+        [Fact]
+        public async Task AcceptGetsCanceledByDispose()
+        {
+            // We try this a couple of times to deal with a timing race: if the Dispose happens
+            // before the operation is started, we won't see a SocketException.
+
+            SocketError? localSocketError = null;
+            bool disposedException = false;
+            for (int i = 0; i < 10 && !localSocketError.HasValue; i++)
+            {
+                var listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+                listener.Listen(1);
+
+                Task acceptTask = Task.Factory.StartNew(() =>
+                {
+                    AcceptAsync(listener).GetAwaiter().GetResult();
+                }, TaskCreationOptions.LongRunning);
+
+                // Wait a little so the operation is started, then Dispose.
+                await Task.Delay(100);
+                Task disposeTask = Task.Factory.StartNew(() =>
+                {
+                    listener.Dispose();
+                }, TaskCreationOptions.LongRunning);
+
+                Task timeoutTask = Task.Delay(30000);
+
+                Assert.NotSame(timeoutTask, await Task.WhenAny(disposeTask, acceptTask, timeoutTask));
```

Comment:
This can instead use the WhenAllOrAnyFailed helper, e.g.
```C#
await new Task[] { acceptTask, disposeTask }.WhenAllOrAnyFailed(millisecondsTimeout: 30_000);
```

Reply:
Ah, you actually don't want to propagate exceptions from the completed task here.  Regardless, if you use Task.Delay for timeouts like this, it's best to cancel them... otherwise they leave pending timers that will eventually fire later, e.g.
```C#
var cts = new CancellationTokenSource();
Task timeoutTask = Task.Delay(30_000, cts.Token);
...
cts.Cancel();
```